### PR TITLE
[CI/Build] Reduce race condition in docker build

### PR DESCRIPTION
**Pull request by youkaichao**
Race condition is observed in https://github.com/vllm-project/vllm/pull/3948 , where docker build sometimes use cached `vllm-nccl` wheel and sometimes use source distribution.

This should be caused by the race condition of cache in docker build, because cache is shared across multiple build.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -16,11 +16,11 @@ RUN ldconfig /usr/local/cuda-12.1/compat/
 
 WORKDIR /workspace
 
-# install build and runtime dependencies
+# copy build and runtime dependency files
+# no need to install now, just copy the files
+# they will be used by `setup.py` to install the dependencies
 COPY requirements-common.txt requirements-common.txt
 COPY requirements-cuda.txt requirements-cuda.txt
```
Comment (simon-mo) [inline Dockerfile:23]: if we are not installing them here then we should not need to copy them

**Comment by simon-mo** [comment]
maybe the right fix is to have separate cache directory per build stage?

**Comment by youkaichao** [comment]
> maybe the right fix is to have separate cache directory per build stage?

How can this solve the problem? I think it is not doable because `setup.py` installs all requirements together, and we cannot let pip use another cache directory for just `vllm-nccl` package.

**Comment by youkaichao** [comment]
didn't see this happen again in months.

**Pull request closed without merging**